### PR TITLE
Parquet: Set parquet bloom filter config with compatible column name

### PR DESCRIPTION
**Pull request by ConeyLiu**
This PR improves the config setting for the parquet bloom filter. 
1. Log warn for those don't exist column or the column type is unsupported (boolean or complex data type).
2. Convert the column name to a compatible one. This is because we make the column name compatible if it contains special character when we write the parquet file.

### CODE REVIEW

**Comment by ConeyLiu** [comment]
Hi, @huaxingao @rdblue @kbendick . Could you help to review this when you are free? Thanks a lot.

File: parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java [hunk 1]
```diff
@@ -231,6 +241,42 @@ private WriteBuilder createContextFunc(
...
+                }
+                Type fieldType = field.type();
+                // doesn't support boolean, list, map, struct type
+                if (!fieldType.isPrimitiveType()
+                    || fieldType.typeId().equals(Type.TypeID.BOOLEAN)) {
+                  LOG.warn(
+                      "The type: {} of column: {} is not supported for parquet bloom filter, ignore it",
+                      fieldType,
+                      columnPath);
+                  return;
```
Comment (kbendick) [inline parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java:267]: Nit / non-blocking question: Is this something we can prevent the user from doing? That might not be worth it though.

Id also format it as `Invalid bloom filter column configuration: field `%s` is of type '%s` which parquet does not have bloom filter support for. This column configuration will be ignored.`. I am a fan of a full constant string at the front for searching logs.

Comment (ConeyLiu) [reply parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java]: Thanks @kbendick for reviewing.

> Is this something we can prevent the user from doing? That might not be worth it though.

I think we may add some checks for adding/updating table properties.

> Id also format it as Invalid bloom filter column configuration: field %s is of type '%s' which cannot benefit from a bloom filter. This column configuration will be ignored.. I am a fan of a full constant string at the front for searching logs.

Updated.

File: parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java [hunk 2]
```diff
@@ -231,6 +241,46 @@ private WriteBuilder createContextFunc(
...
+                      "Invalid bloom filter column configuration: field {} is of type '{}' which cannot "
+                          + "benefit from a bloom filter. This column configuration will be ignored.",
+                      columnPath,
+                      fieldType);
+                  return;
+                }
+
+                String compatibleName =
+                    dotJoiner.join(
+                        Streams.stream(dotSplitter.split(columnPath))
```
Comment (zhongyujiang) [inline parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java:276]: Using dot as splitter seems incorrect to me, cause Iceberg allows dots in field names.

File: parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java [hunk 3]
```diff
@@ -231,6 +241,46 @@ private WriteBuilder createContextFunc(
...
+    private void setBloomFilterConfig(
+        Context context, BiConsumer<String, Boolean> withBloomFilterEnabled) {
+      Splitter dotSplitter = Splitter.on('.');
+      Joiner dotJoiner = Joiner.on(".");
+      context
+          .columnBloomFilterEnabled()
+          .forEach(
+              (columnPath, value) -> {
+                // check the column exists and support bloom filter
+                Types.NestedField field = schema.caseInsensitiveFindField(columnPath);
```
Comment (zhongyujiang) [inline parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java:253]: Maybe we should add this check at the begining, filtering all invalid keys and making field names compatible before pushing these props down to writer. What do you think?

File: parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java [hunk 4]
```diff
@@ -231,6 +241,46 @@ private WriteBuilder createContextFunc(
       return this;
     }
 
+    private void setBloomFilterConfig(
+        Context context, BiConsumer<String, Boolean> withBloomFilterEnabled) {
+      Splitter dotSplitter = Splitter.on('.');
+      Joiner dotJoiner = Joiner.on(".");
```
Comment (rdblue) [inline parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java:247]: Constants should not be created as local variables.

File: parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java [hunk 5]
```diff
@@ -231,6 +241,46 @@ private WriteBuilder createContextFunc(
...
+                      columnPath,
+                      fieldType);
+                  return;
+                }
+
+                String compatibleName =
+                    dotJoiner.join(
+                        Streams.stream(dotSplitter.split(columnPath))
+                            .map(AvroSchemaUtil::makeCompatibleName)
+                            .collect(Collectors.toList()));
```
Comment (rdblue) [inline parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java:278]: This should be separated into a util method in `ParquetSchemaUtil` that can be tested.

File: parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java [hunk 6]
```diff
@@ -231,6 +241,46 @@ private WriteBuilder createContextFunc(
...
+      context
+          .columnBloomFilterEnabled()
+          .forEach(
+              (columnPath, value) -> {
+                // check the column exists and support bloom filter
+                Types.NestedField field = schema.caseInsensitiveFindField(columnPath);
+                if (field == null) {
+                  LOG.warn(
+                      "Invalid bloom filter column configuration, field: {} is not found in the schema: {}, "
+                          + "this column configuration will be ignored.",
```
Comment (rdblue) [inline parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java:257]: This should be `LOG.warn("Skipping bloom filter config for missing field: %s", columnPath);`

File: parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java [hunk 7]
```diff
@@ -231,6 +241,46 @@ private WriteBuilder createContextFunc(
...
+                }
+                Type fieldType = field.type();
+                // doesn't support boolean, list, map, struct type
+                if (!fieldType.isPrimitiveType()
+                    || fieldType.typeId().equals(Type.TypeID.BOOLEAN)) {
+                  LOG.warn(
+                      "Invalid bloom filter column configuration: field {} is of type '{}' which cannot "
+                          + "benefit from a bloom filter. This column configuration will be ignored.",
+                      columnPath,
+                      fieldType);
```
Comment (rdblue) [inline parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java:270]: I don't think this is worth checking.

File: parquet/src/test/java/org/apache/iceberg/parquet/TestBloomRowGroupFilter.java [hunk 8]
```diff
@@ -951,6 +968,72 @@ public void testStructFieldNotEq() {
     }
   }
 
+  @Test
+  public void testStructFieldWithSpecialCharacterLt() {
```
Comment (rdblue) [inline parquet/src/test/java/org/apache/iceberg/parquet/TestBloomRowGroupFilter.java:972]: @huaxingao, can you review these tests?

File: parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java [hunk 9]
```diff
@@ -291,11 +341,7 @@ public <D> FileAppender<D> build() throws IOException {
                 .withMaxRowCountForPageSizeCheck(rowGroupCheckMaxRecordCount)
                 .withMaxBloomFilterBytes(bloomFilterMaxBytes);
 
-        for (Map.Entry<String, String> entry : columnBloomFilterEnabled.entrySet()) {
```
Comment (huaxingao) [inline parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java:294]: nit: you forgot to delete the line
```
Map<String, String> columnBloomFilterEnabled = context.columnBloomFilterEnabled();
```

File: parquet/src/test/java/org/apache/iceberg/parquet/TestBloomRowGroupFilter.java (hunk 8 above)
Comment (huaxingao) [reply parquet/src/test/java/org/apache/iceberg/parquet/TestBloomRowGroupFilter.java:972]: @rdblue These tests look good. Thanks for pinging me.

File: parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java (hunk 2 above)
Comment (ConeyLiu) [reply parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java]: Good catch, updated to use the parquet column path.

File: parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java (hunk 7 above)
Comment (ConeyLiu) [reply parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java:268]: Hi @rdblue, thanks for your time. The most end user doesn't understand the parquet bloom filter implementation details. The log warning here could give some users more information about their settings. Just as @zhongyujiang mentioned, maybe we should add some checking when setting/updating table properties.

**Comment by ConeyLiu** [comment]
Thanks @rdblue @kbendick @zhongyujiang @huaxingao for your time to review. Comments have been addressed. Pls take another look when you are free.

File: parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java (hunk 3 above)
Comment (ConeyLiu) [reply parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java:255]: I think it should do in setting/updating properties. Actually, we already do some simple checkers in our company. I could submit a PR for this and waiting for @rdblue opinion.

**Comment by zinking** [comment]
@huaxingao can you take another look and push forward ? thanks

**Comment by huaxingao** [comment]
LGTM

**Comment by huaxingao** [comment]
cc @rdblue Could you please take another look when you have a moment?

**Pull request closed without merging**